proxy/src/bind.rs

```diff
@@ -323,36 +339,75 @@ where
 
 // ===== impl Binding =====
 
-impl<B: tower_h2::Body + 'static> tower::Service for Binding<B> {
+impl<B: tower_h2::Body + 'static> tower::Service for BoundService<B> {
     type Request = <Stack<B> as tower::Service>::Request;
     type Response = <Stack<B> as tower::Service>::Response;
     type Error = <Stack<B> as tower::Service>::Error;
     type Future = <Stack<B> as tower::Service>::Future;
 
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        match *self {
+        let ready = match self.binding {
             // A service is already bound, so poll its readiness.
             Binding::Bound(ref mut svc) |
-            Binding::BindsPerRequest { next: Some(ref mut svc), .. } => svc.poll_ready(),
+            Binding::BindsPerRequest { next: Some(ref mut svc) } => svc.poll_ready(),
 
             // If no stack has been bound, bind it now so that its readiness can be
             // checked. Store it so it can be consumed to dispatch the next request.
-            Binding::BindsPerRequest { ref endpoint, ref protocol, ref bind, ref mut next } => {
-                let mut svc = bind.bind_stack(endpoint, protocol);
-                let ready = svc.poll_ready()?;
+            Binding::BindsPerRequest { ref mut next } => {
+                let mut svc = self.bind.bind_stack(&self.endpoint, &self.protocol);
+                let ready = svc.poll_ready();
                 *next = Some(svc);
-                Ok(ready)
+                ready
             }
+        };
+
+        // If there was a connect error, don't terminate this BoundService
+        // completely. Instead, simply clean up the inner service, prepare to
+        // make a new one, and tell our caller that we could maybe be ready
+        // if they call `poll_ready` again.
+        //
+        // If they *don't* call `poll_ready` again, that's ok, we won't ever
+        // try to connect again.
+        match ready {
+            Err(ReconnectError::Connect(err)) => {
+                error!("connect error to {:?}: {}", self.endpoint, err);
```

Comment:
While I'd rather more loudly log when connect errors occur, this seems to trigger several times during the test, because of how many times it retries before the metrics scrape noticed the TCP connection event.

This could be reduced from an `error!` to `warn!`, or lower... and/or we could also apply some backoff here so as to not loop back immediately, but after a second or something. (Currently, this yields to the executor, so any other work should be polled again first, and then this will get polled again...)